// @theia/monaco-editor-core/esm/vs/editor/common/languages.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalUriOpenerPriority = exports.TokenizationRegistry = exports.InlayHintKind = exports.CommentMode = exports.CommentThreadState = exports.CommentThreadCollapsibleState = exports.Command = exports.FoldingRangeKind = exports.SymbolKinds = exports.isLocationLink = exports.DocumentHighlightKind = exports.SignatureHelpTriggerKind = exports.InlineCompletionTriggerKind = exports.CompletionItemKinds = exports.EncodedTokenizationResult = exports.TokenizationResult = exports.Token = void 0;
const codicons_1 = require("../../base/common/codicons");
const uri_1 = require("../../base/common/uri");
const range_1 = require("./core/range");
const tokenizationRegistry_1 = require("./tokenizationRegistry");
class Token {
    constructor(offset, type, language) {
        this._tokenBrand = undefined;
        this.offset = offset;
        this.type = type;
        this.language = language;
    }
    toString() {
        return '(' + this.offset + ', ' + this.type + ')';
    }
}
exports.Token = Token;
/**
 * @internal
 */
class TokenizationResult {
    constructor(tokens, endState) {
        this._tokenizationResultBrand = undefined;
        this.tokens = tokens;
        this.endState = endState;
    }
}
exports.TokenizationResult = TokenizationResult;
/**
 * @internal
 */
class EncodedTokenizationResult {
    constructor(tokens, endState) {
        this._encodedTokenizationResultBrand = undefined;
        this.tokens = tokens;
        this.endState = endState;
    }
}
exports.EncodedTokenizationResult = EncodedTokenizationResult;
/**
 * @internal
 */
var CompletionItemKinds;
(function (CompletionItemKinds) {
    const byKind = new Map();
    byKind.set(0 /* CompletionItemKind.Method */, codicons_1.Codicon.symbolMethod);
    byKind.set(1 /* CompletionItemKind.Function */, codicons_1.Codicon.symbolFunction);
    byKind.set(2 /* CompletionItemKind.Constructor */, codicons_1.Codicon.symbolConstructor);
    byKind.set(3 /* CompletionItemKind.Field */, codicons_1.Codicon.symbolField);
    byKind.set(4 /* CompletionItemKind.Variable */, codicons_1.Codicon.symbolVariable);
    byKind.set(5 /* CompletionItemKind.Class */, codicons_1.Codicon.symbolClass);
    byKind.set(6 /* CompletionItemKind.Struct */, codicons_1.Codicon.symbolStruct);
    byKind.set(7 /* CompletionItemKind.Interface */, codicons_1.Codicon.symbolInterface);
    byKind.set(8 /* CompletionItemKind.Module */, codicons_1.Codicon.symbolModule);
    byKind.set(9 /* CompletionItemKind.Property */, codicons_1.Codicon.symbolProperty);
    byKind.set(10 /* CompletionItemKind.Event */, codicons_1.Codicon.symbolEvent);
    byKind.set(11 /* CompletionItemKind.Operator */, codicons_1.Codicon.symbolOperator);
    byKind.set(12 /* CompletionItemKind.Unit */, codicons_1.Codicon.symbolUnit);
    byKind.set(13 /* CompletionItemKind.Value */, codicons_1.Codicon.symbolValue);
    byKind.set(15 /* CompletionItemKind.Enum */, codicons_1.Codicon.symbolEnum);
    byKind.set(14 /* CompletionItemKind.Constant */, codicons_1.Codicon.symbolConstant);
    byKind.set(15 /* CompletionItemKind.Enum */, codicons_1.Codicon.symbolEnum);
    byKind.set(16 /* CompletionItemKind.EnumMember */, codicons_1.Codicon.symbolEnumMember);
    byKind.set(17 /* CompletionItemKind.Keyword */, codicons_1.Codicon.symbolKeyword);
    byKind.set(27 /* CompletionItemKind.Snippet */, codicons_1.Codicon.symbolSnippet);
    byKind.set(18 /* CompletionItemKind.Text */, codicons_1.Codicon.symbolText);
    byKind.set(19 /* CompletionItemKind.Color */, codicons_1.Codicon.symbolColor);
    byKind.set(20 /* CompletionItemKind.File */, codicons_1.Codicon.symbolFile);
    byKind.set(21 /* CompletionItemKind.Reference */, codicons_1.Codicon.symbolReference);
    byKind.set(22 /* CompletionItemKind.Customcolor */, codicons_1.Codicon.symbolCustomColor);
    byKind.set(23 /* CompletionItemKind.Folder */, codicons_1.Codicon.symbolFolder);
    byKind.set(24 /* CompletionItemKind.TypeParameter */, codicons_1.Codicon.symbolTypeParameter);
    byKind.set(25 /* CompletionItemKind.User */, codicons_1.Codicon.account);
    byKind.set(26 /* CompletionItemKind.Issue */, codicons_1.Codicon.issues);
    /**
     * @internal
     */
    function toIcon(kind) {
        let codicon = byKind.get(kind);
        if (!codicon) {
            console.info('No codicon found for CompletionItemKind ' + kind);
            codicon = codicons_1.Codicon.symbolProperty;
        }
        return codicon;
    }
    CompletionItemKinds.toIcon = toIcon;
    const data = new Map();
    data.set('method', 0 /* CompletionItemKind.Method */);
    data.set('function', 1 /* CompletionItemKind.Function */);
    data.set('constructor', 2 /* CompletionItemKind.Constructor */);
    data.set('field', 3 /* CompletionItemKind.Field */);
    data.set('variable', 4 /* CompletionItemKind.Variable */);
    data.set('class', 5 /* CompletionItemKind.Class */);
    data.set('struct', 6 /* CompletionItemKind.Struct */);
    data.set('interface', 7 /* CompletionItemKind.Interface */);
    data.set('module', 8 /* CompletionItemKind.Module */);
    data.set('property', 9 /* CompletionItemKind.Property */);
    data.set('event', 10 /* CompletionItemKind.Event */);
    data.set('operator', 11 /* CompletionItemKind.Operator */);
    data.set('unit', 12 /* CompletionItemKind.Unit */);
    data.set('value', 13 /* CompletionItemKind.Value */);
    data.set('constant', 14 /* CompletionItemKind.Constant */);
    data.set('enum', 15 /* CompletionItemKind.Enum */);
    data.set('enum-member', 16 /* CompletionItemKind.EnumMember */);
    data.set('enumMember', 16 /* CompletionItemKind.EnumMember */);
    data.set('keyword', 17 /* CompletionItemKind.Keyword */);
    data.set('snippet', 27 /* CompletionItemKind.Snippet */);
    data.set('text', 18 /* CompletionItemKind.Text */);
    data.set('color', 19 /* CompletionItemKind.Color */);
    data.set('file', 20 /* CompletionItemKind.File */);
    data.set('reference', 21 /* CompletionItemKind.Reference */);
    data.set('customcolor', 22 /* CompletionItemKind.Customcolor */);
    data.set('folder', 23 /* CompletionItemKind.Folder */);
    data.set('type-parameter', 24 /* CompletionItemKind.TypeParameter */);
    data.set('typeParameter', 24 /* CompletionItemKind.TypeParameter */);
    data.set('account', 25 /* CompletionItemKind.User */);
    data.set('issue', 26 /* CompletionItemKind.Issue */);
    /**
     * @internal
     */
    function fromString(value, strict) {
        let res = data.get(value);
        if (typeof res === 'undefined' && !strict) {
            res = 9 /* CompletionItemKind.Property */;
        }
        return res;
    }
    CompletionItemKinds.fromString = fromString;
})(CompletionItemKinds = exports.CompletionItemKinds || (exports.CompletionItemKinds = {}));
/**
 * How an {@link InlineCompletionsProvider inline completion provider} was triggered.
 */
var InlineCompletionTriggerKind;
(function (InlineCompletionTriggerKind) {
    /**
     * Completion was triggered automatically while editing.
     * It is sufficient to return a single completion item in this case.
     */
    InlineCompletionTriggerKind[InlineCompletionTriggerKind["Automatic"] = 0] = "Automatic";
    /**
     * Completion was triggered explicitly by a user gesture.
     * Return multiple completion items to enable cycling through them.
     */
    InlineCompletionTriggerKind[InlineCompletionTriggerKind["Explicit"] = 1] = "Explicit";
})(InlineCompletionTriggerKind = exports.InlineCompletionTriggerKind || (exports.InlineCompletionTriggerKind = {}));
var SignatureHelpTriggerKind;
(function (SignatureHelpTriggerKind) {
    SignatureHelpTriggerKind[SignatureHelpTriggerKind["Invoke"] = 1] = "Invoke";
    SignatureHelpTriggerKind[SignatureHelpTriggerKind["TriggerCharacter"] = 2] = "TriggerCharacter";
    SignatureHelpTriggerKind[SignatureHelpTriggerKind["ContentChange"] = 3] = "ContentChange";
})(SignatureHelpTriggerKind = exports.SignatureHelpTriggerKind || (exports.SignatureHelpTriggerKind = {}));
/**
 * A document highlight kind.
 */
var DocumentHighlightKind;
(function (DocumentHighlightKind) {
    /**
     * A textual occurrence.
     */
    DocumentHighlightKind[DocumentHighlightKind["Text"] = 0] = "Text";
    /**
     * Read-access of a symbol, like reading a variable.
     */
    DocumentHighlightKind[DocumentHighlightKind["Read"] = 1] = "Read";
    /**
     * Write-access of a symbol, like writing to a variable.
     */
    DocumentHighlightKind[DocumentHighlightKind["Write"] = 2] = "Write";
})(DocumentHighlightKind = exports.DocumentHighlightKind || (exports.DocumentHighlightKind = {}));
/**
 * @internal
 */
function isLocationLink(thing) {
    return thing
        && uri_1.URI.isUri(thing.uri)
        && range_1.Range.isIRange(thing.range)
        && (range_1.Range.isIRange(thing.originSelectionRange) || range_1.Range.isIRange(thing.targetSelectionRange));
}
exports.isLocationLink = isLocationLink;
/**
 * @internal
 */
var SymbolKinds;
(function (SymbolKinds) {
    const byKind = new Map();
    byKind.set(0 /* SymbolKind.File */, codicons_1.Codicon.symbolFile);
    byKind.set(1 /* SymbolKind.Module */, codicons_1.Codicon.symbolModule);
    byKind.set(2 /* SymbolKind.Namespace */, codicons_1.Codicon.symbolNamespace);
    byKind.set(3 /* SymbolKind.Package */, codicons_1.Codicon.symbolPackage);
    byKind.set(4 /* SymbolKind.Class */, codicons_1.Codicon.symbolClass);
    byKind.set(5 /* SymbolKind.Method */, codicons_1.Codicon.symbolMethod);
    byKind.set(6 /* SymbolKind.Property */, codicons_1.Codicon.symbolProperty);
    byKind.set(7 /* SymbolKind.Field */, codicons_1.Codicon.symbolField);
    byKind.set(8 /* SymbolKind.Constructor */, codicons_1.Codicon.symbolConstructor);
    byKind.set(9 /* SymbolKind.Enum */, codicons_1.Codicon.symbolEnum);
    byKind.set(10 /* SymbolKind.Interface */, codicons_1.Codicon.symbolInterface);
    byKind.set(11 /* SymbolKind.Function */, codicons_1.Codicon.symbolFunction);
    byKind.set(12 /* SymbolKind.Variable */, codicons_1.Codicon.symbolVariable);
    byKind.set(13 /* SymbolKind.Constant */, codicons_1.Codicon.symbolConstant);
    byKind.set(14 /* SymbolKind.String */, codicons_1.Codicon.symbolString);
    byKind.set(15 /* SymbolKind.Number */, codicons_1.Codicon.symbolNumber);
    byKind.set(16 /* SymbolKind.Boolean */, codicons_1.Codicon.symbolBoolean);
    byKind.set(17 /* SymbolKind.Array */, codicons_1.Codicon.symbolArray);
    byKind.set(18 /* SymbolKind.Object */, codicons_1.Codicon.symbolObject);
    byKind.set(19 /* SymbolKind.Key */, codicons_1.Codicon.symbolKey);
    byKind.set(20 /* SymbolKind.Null */, codicons_1.Codicon.symbolNull);
    byKind.set(21 /* SymbolKind.EnumMember */, codicons_1.Codicon.symbolEnumMember);
    byKind.set(22 /* SymbolKind.Struct */, codicons_1.Codicon.symbolStruct);
    byKind.set(23 /* SymbolKind.Event */, codicons_1.Codicon.symbolEvent);
    byKind.set(24 /* SymbolKind.Operator */, codicons_1.Codicon.symbolOperator);
    byKind.set(25 /* SymbolKind.TypeParameter */, codicons_1.Codicon.symbolTypeParameter);
    /**
     * @internal
     */
    function toIcon(kind) {
        let icon = byKind.get(kind);
        if (!icon) {
            console.info('No codicon found for SymbolKind ' + kind);
            icon = codicons_1.Codicon.symbolProperty;
        }
        return icon;
    }
    SymbolKinds.toIcon = toIcon;
})(SymbolKinds = exports.SymbolKinds || (exports.SymbolKinds = {}));
class FoldingRangeKind {
    /**
     * Creates a new {@link FoldingRangeKind}.
     *
     * @param value of the kind.
     */
    constructor(value) {
        this.value = value;
    }
}
exports.FoldingRangeKind = FoldingRangeKind;
/**
 * Kind for folding range representing a comment. The value of the kind is 'comment'.
 */
FoldingRangeKind.Comment = new FoldingRangeKind('comment');
/**
 * Kind for folding range representing a import. The value of the kind is 'imports'.
 */
FoldingRangeKind.Imports = new FoldingRangeKind('imports');
/**
 * Kind for folding range representing regions (for example marked by `#region`, `#endregion`).
 * The value of the kind is 'region'.
 */
FoldingRangeKind.Region = new FoldingRangeKind('region');
/**
 * @internal
 */
var Command;
(function (Command) {
    /**
     * @internal
     */
    function is(obj) {
        if (!obj || typeof obj !== 'object') {
            return false;
        }
        return typeof obj.id === 'string' &&
            typeof obj.title === 'string';
    }
    Command.is = is;
})(Command = exports.Command || (exports.Command = {}));
/**
 * @internal
 */
var CommentThreadCollapsibleState;
(function (CommentThreadCollapsibleState) {
    /**
     * Determines an item is collapsed
     */
    CommentThreadCollapsibleState[CommentThreadCollapsibleState["Collapsed"] = 0] = "Collapsed";
    /**
     * Determines an item is expanded
     */
    CommentThreadCollapsibleState[CommentThreadCollapsibleState["Expanded"] = 1] = "Expanded";
})(CommentThreadCollapsibleState = exports.CommentThreadCollapsibleState || (exports.CommentThreadCollapsibleState = {}));
/**
 * @internal
 */
var CommentThreadState;
(function (CommentThreadState) {
    CommentThreadState[CommentThreadState["Unresolved"] = 0] = "Unresolved";
    CommentThreadState[CommentThreadState["Resolved"] = 1] = "Resolved";
})(CommentThreadState = exports.CommentThreadState || (exports.CommentThreadState = {}));
/**
 * @internal
 */
var CommentMode;
(function (CommentMode) {
    CommentMode[CommentMode["Editing"] = 0] = "Editing";
    CommentMode[CommentMode["Preview"] = 1] = "Preview";
})(CommentMode = exports.CommentMode || (exports.CommentMode = {}));
var InlayHintKind;
(function (InlayHintKind) {
    InlayHintKind[InlayHintKind["Type"] = 1] = "Type";
    InlayHintKind[InlayHintKind["Parameter"] = 2] = "Parameter";
})(InlayHintKind = exports.InlayHintKind || (exports.InlayHintKind = {}));
/**
 * @internal
 */
exports.TokenizationRegistry = new tokenizationRegistry_1.TokenizationRegistry();
/**
 * @internal
 */
var ExternalUriOpenerPriority;
(function (ExternalUriOpenerPriority) {
    ExternalUriOpenerPriority[ExternalUriOpenerPriority["None"] = 0] = "None";
    ExternalUriOpenerPriority[ExternalUriOpenerPriority["Option"] = 1] = "Option";
    ExternalUriOpenerPriority[ExternalUriOpenerPriority["Default"] = 2] = "Default";
    ExternalUriOpenerPriority[ExternalUriOpenerPriority["Preferred"] = 3] = "Preferred";
})(ExternalUriOpenerPriority = exports.ExternalUriOpenerPriority || (exports.ExternalUriOpenerPriority = {}));
